Tutorial for creating a dataset (#5540)
huggingface/datasets: docs/source/create_dataset.mdx

```

Any additional information about your dataset, such as text captions or transcriptions, can be included with a `metadata.csv` file in the folder containing your dataset. The metadata file needs to have a `file_name` column that links the image or audio file to its corresponding metadata:

```
file_name, text
bulbasaur.png, There is a plant seed on its back right from the day this Pokémon is born.
charmander.png, It has a preference for hot things.
squirtle.png, When it retracts its long neck into its shell, it squirts out water with vigorous force.
```

To learn more about each of these folder-based builders, check out the and <a href="https://huggingface.co/docs/datasets/image_dataset#imagefolder"><span class="underline decoration-yellow-400 decoration-2 font-semibold">ImageFolder</span></a> or <a href="https://huggingface.co/docs/datasets/audio_dataset#audiofolder"><span class="underline decoration-pink-400 decoration-2 font-semibold">AudioFolder</span></a> guides.

## From local files

You can also create a dataset from local files by specifying the path to the data files. There are two ways you can create a dataset using the `from_` methods:

>     * The [`~Dataset.from_generator`] method is the most memory-efficient way to create a dataset from a [generator](https://wiki.python.org/moin/Generators) due to a generators lazy behavior. This is especially useful when you"re working with a really large dataset that may not fit in memory.

Reviewer: I would keep the work "lazy" for iterable datasets

```suggestion
    * The [`~Dataset.from_generator`] method is the most memory-efficient way to create a dataset from a [generator](https://wiki.python.org/moin/Generators) due to a generators iterative behavior. This is especially useful when you"re working with a really large dataset that may not fit in memory, since the dataset is generated on disk progressively and then memory-mapped.
```

    ```py
    >>> from datasets import Dataset
    >>> def gen():
    ...     yield {"pokemon": "bulbasaur", "type": "grass"}
    ...     yield {"pokemon": "squirtle", "type": "water"}

>     >>> ds = Dataset.from_generator(gen)
>     ```
> 
>     Generator-based datasets need to be iterated over with either a `for` loop or by calling `next(iter())`:
> 
>     ```py
>     >>> for example in ds:
>     ...     print(example)
>     {"pokemon": "bulbasaur", "type": "grass"}
>     {"pokemon": "squirtle", "type": "water"}

Reviewer: A Dataset created with a generator can be queried using normal indexing:
```suggestion
    >>> ds = Dataset.from_generator(gen)
    >>> ds[0]
    {"pokemon": "bulbasaur", "type": "grass"}
    >>> ds[1]
    {"pokemon": "squirtle", "type": "water"}
```

Not sure if it fits the tutorial scope here, but we can also mention `IterableDataset` that indeed requires to be iterated with a `for` loop:

```python
    >>> ds = IterableDataset.from_generator(gen)
    >>> for example in ds:
    ...     print(example)
    {"pokemon": "bulbasaur", "type": "grass"}
    {"pokemon": "squirtle", "type": "water"}
```

## From local files

You can also create a dataset from local files by specifying the path to the data files. There are two ways you can create a dataset using the `from_` methods:

    * The [`~Dataset.from_generator`] method is the most memory-efficient way to create a dataset from a [generator](https://wiki.python.org/moin/Generators) due to a generators iterative behavior. This is especially useful when you're working with a really large dataset that may not fit in memory, since the dataset is generated on disk progressively and then memory-mapped.

    ```py
    >>> from datasets import Dataset
    >>> def gen():
    ...     yield {"pokemon": "bulbasaur", "type": "grass"}
    ...     yield {"pokemon": "squirtle", "type": "water"}
    >>> ds = Dataset.from_generator(gen)
    >>> ds[0]
    {"pokemon": "bulbasaur", "type": "grass"}
    ```

    A generator-based [`IterableDataset`] needs to be iterated over with a `for` loop for example:

    ```py

>     >>> ds = IterableDataset.from_generator(gen)

Reviewer: ```suggestion
    >>> from datasets import IterableDataset
    >>> ds = IterableDataset.from_generator(gen)
```
